Remove react-bootstrap-toggle dependency and replace ReactBootstrapToggle usages

cnathe opened the pull request:
#### Rationale
Selenium test update for ToggleButton conversion from ReactBootstrapToggle to ToggleButtons/ToggleIcon

#### Related Pull Requests
* https://github.com/LabKey/labkey-ui-components/pull/1326

#### Changes
- update ToggleButton to account for button or icon version of ToggleButtons.tsx component

cnathe committed:
Selenium test update for ToggleButton conversion from ReactBootstrapToggle to ToggleButtons/ToggleIcon

labkey-tchad commented on src/org/labkey/test/components/domain/DomainFormPanel.java:
```diff
@@ -394,21 +394,20 @@ public boolean isDefaultSystemFieldsExpanded()
     public String getMode()
     {
-        return elementCache().customFieldsViewToggle.getSelectedStatus(); // will be either "Summary mode" or "Detail mode"
+        return elementCache().customFieldsViewToggle.getSelectedStatus(); // will be either "Summary" or "Detail"
```
I know this isn't open for review yet but I have some feedback since you're refactoring these anyway. Tests shouldn't have to know the specific toggle text for getting or setting the mode.
Outside of this class, the mode could be represented by an `enum` (e.g. `public enum DomainFormMode { Summary, Detail }`) which the public setter and getter would use.
Another option would be to get rid of (or make private) the existing `getMode`/`switchMode` methods and restrict callers to using `isDetailMode`/`isSummaryMode`/`selectDetailMode`/`selectSummaryMode`.

cnathe replied:
Good suggestion. I'll look into making those changes and ping you for review once ready.

labkey-nicka replied:
Leaving this PR review for @labkey-tchad.

cnathe replied:
Done

cnathe committed:
Fix for dataset/list designer import from file toggle

cnathe committed:
DomainFormPanel update to add selectSummaryMode() and selectDetailMode() helpers
- and make getMode() and switchMode() private

labkey-tchad commented on src/org/labkey/test/components/react/ToggleButton.java:
```diff
@@ -32,31 +32,57 @@ public WebDriver getDriver()
+            if (hasButtons()) selectFirst();
+            else getComponentElement().click();
+        }
+        else if (!enabled && isEnabled())
+        {
+            if (hasButtons()) selectSecond();
+            else getComponentElement().click();
+        }
```
This should have some failure state if `isEnabled` and `isDisabled` are both false. That would indicate that the test component is out of sync with the product, right?
```suggestion
        else if (!enabled && isEnabled())
        {
            if (hasButtons()) selectSecond();
            else getComponentElement().click();
        }
        else
        {
            throw new IllegalStateException("Unable to determine toggle state: " + getComponentElement());
        }
```

cnathe replied:
Yes. Added.

labkey-tchad commented on src/org/labkey/test/components/react/ToggleButton.java:
```diff
@@ -32,31 +32,57 @@ public WebDriver getDriver()
+    {
+        Locator.tag("button").index(1).findElement(this).click();
+        return this;
+    }
```
Do `selectFirst`/`selectSecond` need to be public? They seem to complicate the component.

cnathe replied:
nope, good catch. updated.

labkey-tchad commented on src/org/labkey/test/components/react/ToggleButton.java:
```diff
@@ -32,31 +32,57 @@ public WebDriver getDriver()
+        return this;
+    }
+
+    private boolean hasButtons()
+    {
+        return Locator.tag("button").findElementOrNull(this) != null;
```
I would stash the `hasButtons` value once to avoid searching for the buttons with every operation.
```suggestion
    private Boolean hasButtons = null;
    private boolean hasButtons()
    {
        if (hasButtons == null)
            hasButtons = Locator.tag("button").existsIn(this);
        return hasButtons;
```

cnathe replied:
good idea. done

cnathe committed:
Merge branch 'develop' into fb_removeReactBootstrapToggle

labkey-tchad commented on src/org/labkey/test/components/react/ToggleButton.java:
```diff
@@ -32,31 +32,57 @@ public WebDriver getDriver()
+
+    public boolean isDisabled()
+    {
+        return getComponentElement().getAttribute("class").contains("toggle-off");
```
Why was `get` split into two methods? I would expect these to be symmetrical and thus redundant to have them both.
Also, `isEnabled` is misleading. In other contexts, that would indicate whether or not one would be allowed to toggle the control (i.e. the `disabled` attribute), not what the on/off state is.

cnathe replied:
Good points. I removed `isDisabled` and renamed `isEnabled` to `isOn`

cnathe committed:
Test CR feedback
- add error case to ToggleButton.set() and remove extra calls to setEditableState() from usages
- rename isEnabled() to isOn() and remove isDisabled()
- make serveral methods form ToggleButton private

cnathe committed:
EditTemplatePage refactor to use ToggleButton test helper

cnathe committed:
Merge branch 'develop' into fb_removeReactBootstrapToggle
